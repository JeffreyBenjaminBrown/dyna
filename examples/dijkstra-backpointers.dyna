% single source shortest path with optimal path extraction.
path(start) argmin= [0, start].
path(B) argmin= [path(A) + edge(A,B), A].
goal argmin= [path(end), end].

% expensive path
edge("a","b") := 1.
edge("b","c") := 1.
edge("c","d") := 1.

% cheap path
edge("a","e") := 1.
edge("e","d") := 1.

% define begining and end.
start := "a".
end := "d".

% extract cheapest path by following backpointers from each vertex.
bestpath(start) := [start].
bestpath(V) := U is $key(&path(V)), [V | bestpath(U)] for V != start.

% the optimal path is the one from the `end`.
optimalpath = reverse(bestpath(end)).

% ---------------------------------------
% list utilities
:- backchain reverse/1.
reverse([]) = [].
reverse([X|Xs]) = append(reverse(Xs), X).

:- backchain append/2.
append([], Y) = [Y].
append([X|Xs], Y) = [X|append(Xs, Y)].
