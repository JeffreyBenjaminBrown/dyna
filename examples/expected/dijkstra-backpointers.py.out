
Solution
========
end = "d".
goal = 2.
optimalpath = ["a", "e", "d"].
start = "a".

$key/1
======
$key(goal)      = end.
$key(path("a")) = start.
$key(path("b")) = "a".
$key(path("c")) = "b".
$key(path("d")) = "e".
$key(path("e")) = "a".

bestpath/1
==========
bestpath("a") = ["a"].
bestpath("b") = ["b", "a"].
bestpath("c") = ["c", "b", "a"].
bestpath("d") = ["d", "e", "a"].
bestpath("e") = ["e", "a"].

edge/2
======
edge("a","b") = 1.
edge("a","e") = 1.
edge("b","c") = 1.
edge("c","d") = 1.
edge("e","d") = 1.

path/1
======
path("a") = 0.
path("b") = 1.
path("c") = 2.
path("d") = 2.
path("e") = 1.

